Prom histograms in Flux draft-in-progress (#2525)
influxdata/docs-v2: content/influxdb/v2.0/query-data/flux/histograms.md

4. Reshape the data so all duration buckets for the same period are in their own tables, with an upper bound column that describes the bucket represented by each row.
5. Transform each table from a histogram to a quantile with the `histogramQuantile()` function.

The following query performs the above steps.

```js
import "experimental"

// The "_field" is necessary. Any columns following "_field" will be used
// to create quantiles for each unique value in that column.
// E.g., put "path" in here to see quantiles for each unique value of "path".
groupCols = ["_field"]

// This is a helper function that takes a stream of tables,
// each containing "le" buckets for one window period.
// It uses histogramQuantile to transform the bin counts
// to a quantile defined in the "q" parameter.
// The windows are then reassembled to produce a single table for each
// unique group key.
doQuantile = (tables=<-, q) => tables
  |> histogramQuantile(quantile: q)
  |> duplicate(as: "_time", column: "_stop")
  |> window(every: inf)
  |> map(fn: (r) => ({r with _measurement: "quantile", _field: string(v: q)}))
  |> experimental.group(mode: "extend", columns: ["_measurement", "_field"])

> histos =

Reviewer: ```suggestion
histograms =
```

  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r._measurement == "http_api_request_duration_seconds"
    and r._field != "count" and r._field != "sum")
  // Downsample the data. This helps a lot with performance!
  |> aggregateWindow(fn: last, every: v.windowPeriod)
  // Calling difference() transforms the cumulative count of requests
  // to the number of new requests per window period.
  |> difference(nonNegative: true)
  // Counters may be reset when a server restarts.
  // When this happens there will be null values produced by difference().
  |> filter(fn: (r) => exists r._value)
  // Group data on the requested dimensions, window it, and sum within those dimensions, for each window.
  |> group(columns: groupCols)
  |> window(every: v.windowPeriod)
  |> sum()
  // Fields will have names like "0.001", etc. Change _field to a float column called "le".
  // This also has the effect of ungrouping by _field, so bucket counts for each period
  // will be within the same table.
  |> map(fn: (r) => ({r with le: float(v: r._field)}))
  |> drop(columns: ["_field"])

// Compute the 50th and 95th percentile for duration of http requests.
union(tables: [

>   histos |> doQuantile(q:  0.95),
>   histos |> doQuantile(q:  0.5)

Reviewer: ```suggestion
  histograms |> doQuantile(q:  0.95),
  histograms |> doQuantile(q:  0.5)
```

])
```

> The tables flowing into the call to `histogramQuantile` should look similar to this. Note, that rows are sorted by `le` to be clear that the counts increase for larger upper bounds):

Reviewer: ```suggestion
The tables piped-forward into `histogramQuantile()` should look similar to those returned by the `histograms` variable in the example above. Note, that rows are sorted by `le` to be clear that the counts increase for larger upper bounds.
```
